File: javascript/88.merge-sorted-array.js
/*
 * @lc app=leetcode id=88 lang=javascript
 *
 * [88] Merge Sorted Array
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function (nums1, m, nums2, n) {
  // Solution 1:
  // for (let i = 0; i < n; i++) {
  //   nums1[m + i] = nums2[i];
  // }
  // // can use any sort methods
  // nums1.sort((a, b) => a - b);

  // Solution 2: two pointers
  let i = m + n - 1,
    p = m - 1,
    q = n - 1;
  while (q >= 0) {
    nums1[i--] = p >= 0 && nums1[p] > nums2[q] ? nums1[p--] : nums2[q--];
  }
  
  return nums1;
};
// @lc code=end

File: rust/88.merge-sorted-array.rs
/*
 * @lc app=leetcode id=88 lang=rust
 *
 * [88] Merge Sorted Array
 */

// @lc code=start
impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
        for i in 0..n {
            nums1[(m + i) as usize] = nums2[i as usize];
        }
        nums1.sort_unstable();

        // let (mut i, mut j, mut k, mut d): (usize, usize, usize, usize) = (0, 0, 0, 0);
        // while (j as i32) < m && (k as i32) < n {
        //     if nums2[k] < nums1[j + d] {
        //         nums1.insert(i, nums2[k]);
        //         nums1.pop();
        //         k += 1;
        //         d += 1;
        //     } else {
        //         j += 1
        //     }
        //     i += 1;
        // }
        // while (k as i32) < n {
        //     nums1[i] = nums2[k];
        //     i += 1;
        //     k += 1;
        // }
    }
}
// @lc code=end
